flocker/control/test/test_httpapi.py

```diff
@@ -450,3 +478,299 @@ def test_listens_endpoint(self):
+                                                image=object(),
+                                                volume=volume)}),
+        )
+        expected_manifestation = Manifestation(dataset=expected_dataset,
+                                               primary=False)
+        node2 = Node(
+            hostname=u"node2.example.com",
+            applications=frozenset(),
+            other_manifestations=frozenset([expected_manifestation])
+        )
+
+        deployment = Deployment(nodes=frozenset([node1, node2]))
+        expected = dict(
+            dataset_id=expected_dataset.dataset_id,
+            primary=expected_hostname,
+            metadata=thaw(expected_dataset.metadata)
+        )
+        self.assertEqual(
+            [expected], list(datasets_from_deployment(deployment)))
+
+    def test_replica_manifestations_only(self):
+        """
+        ``datasets_from_deployment`` does not return datasets if there are only
+        replica manifestations.
+        """
+        manifestation1 = Manifestation(
+            dataset=Dataset(dataset_id=unicode(uuid4())),
+            primary=False
+        )
+        manifestation2 = Manifestation(
+            dataset=Dataset(dataset_id=unicode(uuid4())),
+            primary=False
+        )
+
+        node1 = Node(
+            hostname=u"node1.example.com",
+            applications=frozenset(),
+            other_manifestations=frozenset([manifestation1])
+        )
+
+        node2 = Node(
+            hostname=u"node2.example.com",
+            applications=frozenset(),
+            other_manifestations=frozenset([manifestation2])
+        )
+
+        deployment = Deployment(nodes=frozenset([node1, node2]))
+
+        self.assertEqual([], list(datasets_from_deployment(deployment)))
+
+    def test_multiple_primary_manifestations(self):
+        """
+        ``datasets_from_deployment`` may return multiple primary datasets.
+        """
+        manifestation1 = Manifestation(
+            dataset=Dataset(dataset_id=unicode(uuid4())),
+            primary=True
+        )
+        manifestation2 = Manifestation(
+            dataset=Dataset(dataset_id=unicode(uuid4())),
+            primary=True
+        )
+
+        node1 = Node(
+            hostname=u"node1.example.com",
+            applications=frozenset(),
+            other_manifestations=frozenset([manifestation1])
+        )
+
+        node2 = Node(
+            hostname=u"node2.example.com",
+            applications=frozenset(),
+            other_manifestations=frozenset([manifestation2])
+        )
+
+        deployment = Deployment(nodes=frozenset([node1, node2]))
+        self.assertEqual(
+            set([manifestation1.dataset.dataset_id,
+                 manifestation2.dataset.dataset_id]),
+            set(d['dataset_id'] for d in datasets_from_deployment(deployment))
+        )
+
+
+class ApiDatasetFromDatasetAndNodeTests(SynchronousTestCase):
```

Comment:
Acronyms should be all upper case, even for a CamelCase identifier.  `APIDatasetFrom...`.